### PR TITLE
[IMP] mail: turn model props to model fields

### PR DESCRIPTION
**Pull request by alexkuhn**
Task-2734101

### CODE REVIEW

File: addons/mail/static/src/models/chatter/chatter.js [hunk 1]
```diff
@@ -254,22 +250,37 @@ registerModel({
...
-                this._isPreparingAttachmentsLoading = false;
-            }, this.messaging.loadingBaseDelayDuration);
+            this.update({
+                _attachmentsLoaderTimeout: this.env.browser.setTimeout(
+                    () => {
+                        this.update({
+                            isShowingAttachmentsLoading: true,
+                            _isPreparingAttachmentsLoading: clear(),
+                        });
+                    },
```
Comment (seb-odoo) [inline addons/mail/static/src/models/chatter/chatter.js:260]: general in whole diff: let's take the opportunity to create methods on the model rather than nested anonymousgeneral in whole diff: let's take the opportunity to create methods on the model rather than nested anonymous

File: addons/mail/static/src/models/chatter/chatter.js [hunk 2]
```diff
@@ -254,22 +250,37 @@ registerModel({
...
-            this._attachmentsLoaderTimeout = null;
-            this._isPreparingAttachmentsLoading = false;
+            this.update({
+                _attachmentsLoaderTimeout: clear(),
+                _isPreparingAttachmentsLoading: clear(),
+            });
         },
     },
     fields: {
+        _attachmentsLoaderTimeout: attr({
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/models/chatter/chatter.js:278]: Just kept the exact same name so the change is easier to digest. But I have no desire to introduce private/public fields. Field names are just names.

File: addons/mail/static/src/models/chatter/chatter.js (hunk 1 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/chatter/chatter.js]: Made change for all but the 2 record methods `_debounce` in `RtcCallViewer` and `RtcSession`, because they call function `f` passed as param so we cannot just pass the function with no context.

File: addons/mail/static/src/models/chatter/chatter.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/chatter/chatter.js:282]: Then I would say we should rename them immediately because this is confusing.

Also this will highlight them on the diff, which will help make sure we didn't forget to adapt update() for any of them.

File: addons/mail/static/src/models/chatter/chatter.js [hunk 3]
```diff
@@ -254,22 +259,32 @@ registerModel({
...
+            default: null,
```
Comment (seb-odoo) [inline addons/mail/static/src/models/chatter/chatter.js:283]: there is a need for `null` instead of `undefined`?

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js [hunk 4]
```diff
@@ -175,7 +175,6 @@ registerModel({
         thread: many2one('Thread', {
             compute: '_computeThread',
             readonly: true,
-            required: true,
```
Comment (seb-odoo) [inline addons/mail/static/src/models/file_uploader_view/file_uploader_view.js:178]: why this change?

File: addons/mail/static/src/models/messaging/messaging.js [hunk 5]
```diff
@@ -369,7 +369,6 @@ registerModel({
             default: insertAndReplace(),
             isCausal: true,
             readonly: true,
-            required: true,
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging/messaging.js:372]: why this change?

File: addons/mail/static/src/models/rtc_session/rtc_session.js [hunk 6]
```diff
@@ -241,12 +238,17 @@ registerModel({
...
-                f();
-            }, delay);
+            this.update({
+                _timeoutId: browser.setTimeout(
+                    () => {
+                        if (!this.exists()) {
+                            return;
+                        }
+                        f();
+                    },
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_session/rtc_session.js:248]: I would still make a method for those, it's possible to pass the function as param to the method with `.bind(this, f)` when registering the timeout

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js [hunk 7]
```diff
@@ -49,7 +49,7 @@ registerModel({
             if (this.activityView) {
                 return replace(this.activityView.activity.thread);
             }
-            if (this.attachmentBoxView) {
+            if (this.attachmentBoxView && this.attachmentBoxView.chatter.thread) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/file_uploader_view/file_uploader_view.js:52]: why change?

File: addons/mail/static/src/models/messaging/messaging.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/messaging/messaging.js:372]: I had actual crash in some tests of `messaging.time` being `undefined`

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js:178]: I had failure in tests that `_computeThread` made `replace(undefined)` with `Chatter` as owner.
`Chatter/thread` is not necessarily set on `chatter`, so I removed the `required` because it's wrong in this case.

File: addons/mail/static/src/models/chatter/chatter.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/chatter/chatter.js:283]: Just to keep same values as before turning them to fields. But we can use `undefined` instead of `null`.

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js (hunk 7 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js:52]: Same reason as with https://github.com/odoo/odoo/pull/82698#discussion_r784352069
Had actual crash in tests and turns out `Chatter/thread` can be unset, so we have to guard it otherwise it does `replace(undefined)` which raises an error.

File: addons/mail/static/src/models/messaging/messaging.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/messaging/messaging.js:372]: I had this crash before `Chatter/thread` and fixing it, so I'll check again whether I still have the crash, and if that's still the case investigate what's going on

Comment (alexkuhn) [reply addons/mail/static/src/models/messaging/messaging.js:372]: Checked again and test crashes during teardown with `messaging.time` being undefined.

Apparently during teardown, we got to a point in which `messaging` exists but `messaging.time` no longer exists and there is no crash. `Chatter/_stopAttachmentsLoading` now triggers an update cycle due to props with attachment loading and timeout being turned into field. This is the actual update cycle that detects missing required on `messaging.time` and make the crash. Note that `Chatter/_stopAttachmentsLoading` is called in `Chatter/_willDelete`, so this is a nested update cycle.

Here I just removed the `required` on `Messaging/time` to allow same scenario as before with no crash with introduction of this update cycle that makes the check.

I tried keeping `required: true` on `Messaging/time` and not trigger update cycle from `Chatter/_stopAttachmentsLoading` (just kept `clearTimeout`), and tests pass successfully. I just pushed this version on this PR.

I think all of this shows that nested update cycles can be make checks on state, which is a problem because state of modelling makes sense before and after a full update cycle, not in-between.

File: addons/mail/static/src/models/chatter/chatter.js [hunk 8]
```diff
@@ -25,12 +25,8 @@ registerModel({
     name: 'Chatter',
     identifyingFields: ['id'],
     lifecycleHooks: {
-        _created() {
-            this._attachmentsLoaderTimeout = undefined;
-            this._isPreparingAttachmentsLoading = undefined;
-        },
         _willDelete() {
-            this._stopAttachmentsLoading();
+            this.env.browser.clearTimeout(this.attachmentsLoaderTimeout);
```
Comment (seb-odoo) [inline addons/mail/static/src/models/chatter/chatter.js:29]: Any specific reason to no longer clear the fields on delete? Just because it is kinda useless?

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js]: What kind of chatter doesn't have a thread suddenly with this PR? This is a mistake from the test then, we shouldn't need to change the code.

File: addons/mail/static/src/models/user_setting/user_setting.js [hunk 9]
```diff
@@ -179,8 +178,8 @@ registerModel({
          * @param {any} key
          */
         _debounce(f, delay, key) {
-            this._timeoutIds[key] && browser.clearTimeout(this._timeoutIds[key]);
-            this._timeoutIds[key] = browser.setTimeout(() => {
+            this.timeoutIds[key] && browser.clearTimeout(this.timeoutIds[key]);
+            this.timeoutIds[key] = browser.setTimeout(() => {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/user_setting/user_setting.js:182]: It probably doesn't make a difference here, but writing on a field without `update()` shouldn't be done.

File: addons/mail/static/src/models/chatter/chatter.js (hunk 8 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/chatter/chatter.js:29]: It's related to last paragraph of this comment:
https://github.com/odoo/odoo/pull/82698#discussion_r784748266

to sum up, `.update()` from `_willDelete` triggers a nested update cycle during teardown of tests, and so checks in this nested update cycle make a crash because state is inconsistent.

I have removed this useless nested cycle to fix the issue

File: addons/mail/static/src/models/user_setting/user_setting.js (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/user_setting/user_setting.js:182]: Agree, I just turn into fields in a straight-forward way.
Could introduce a `Dict` model, but I felt it's out of scope of PR and good improvement at later date,

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js]: Ok, will do the change to enforce chatter having a thread. I suppose we didn't set it `required` and have conditional on `Chatter/thread` due to how we dealt with temporary threads or waiting for `init_messaging` RPC on page load to have a thread... But now we know JS modelling is designed with incomplete knowledge so it makes sense to have `Chatter/thread` relation immediately even though info on `thread` could be lacking.

File: addons/mail/static/src/models/user_setting/user_setting.js (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/user_setting/user_setting.js:182]: will do that then

File: addons/mail/static/src/models/file_uploader_view/file_uploader_view.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js]: If we model "temporary thread" chatter differently (eg. by having empty thread), then yes it could become not required, but to do that correctly imply a lot of changes in the whole code of chatter and related components and models, so way out of scope here.

Comment (alexkuhn) [reply addons/mail/static/src/models/file_uploader_view/file_uploader_view.js]: Hm... looks like I didn't have to do anything to make test pass. I see that insert of `Chatter/thread` come from onChange `_onThreadIdOrThreadModelChanged` so possibly had the issue due to nested update cycle that I just removed...

I won't risk killing this onChange in this PR with all current changes. It works at the moment so it's safe to not touch it for the time being.

File: addons/mail/static/src/models/user_setting/user_setting.js (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/user_setting/user_setting.js]: Just did that for all `Object` attributes introduced in this PR (and the `Set` has been turned into `Object` attribute).
Not super pretty with spread object for update (and deletion of item in 3 statements), but changes on these fields are now observed.

**Comment by alexkuhn** [comment]
Just made a PR for each prop turned into field (see task for list of PRs)

**Comment by seb-odoo** [comment]
obsolete pr
